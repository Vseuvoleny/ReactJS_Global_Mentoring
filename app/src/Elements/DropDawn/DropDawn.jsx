import React, { useRef, useEffect } from "react";
import CloseIcon from "../Icon/Icons/CloseIcon/CloseIcon";
import PropTypes from "prop-types";
import "./DropDawn.scss";
const DropDawn = ({ setDropDawnOpen, dispatch }) => {
  const ref = useRef(null);
  const svgRef = useRef(null);

  const outSideClick = (e) => {
    if (
      ref.current &&
      ref.current.contains(e.target) &&
      !svgRef.current.contains(e.target)
    ) {
    } else {
      setDropDawnOpen(false);
    }
  };
  useEffect(() => {
    document.addEventListener("click", outSideClick, true);
    return () => {
      document.removeEventListener("click", outSideClick, true);
    };
  });
  return (
    <div className="dropdawn__container" ref={ref}>
      <CloseIcon
        width="12"
        height="13"
        viewBox="0 0 12 13"
        classes={"close_icon"}
        onClick={outSideClick}
        ref={svgRef}
      />
      <div
        className="control"
        onClick={() => {
          dispatch({ type: "edit" });
        }}
      >
        Edit
      </div>
      <div
        className="control"
        onClick={() => {
          dispatch({ type: "delete" });
        }}
      >
        Delete
      </div>
    </div>
  );
};

DropDawn.propTypes = {
  setDropDawnOpen: PropTypes.func,
  dispatch: PropTypes.func,
};

export default DropDawn;
