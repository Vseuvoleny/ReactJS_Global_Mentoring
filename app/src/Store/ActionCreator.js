import {
  GET_FILMS_ASYNC,
  SET_FILTER,
  SET_GENRES,
  FILM_REQUEST_FAILURE,
} from "./ActionTypes";

export const getFilms = (payload) => ({ type: GET_FILMS_ASYNC, payload });
export const setGenres = (payload) => ({ type: SET_GENRES, payload });
export const setFilter = (payload) => ({ type: SET_FILTER, payload });
export const getErrorFromFetch = (error) => ({
  type: FILM_REQUEST_FAILURE,
  payload: { error },
});
