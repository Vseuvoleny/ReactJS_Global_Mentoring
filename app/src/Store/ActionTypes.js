export const APPLY_MODAL = "APPLY_MODAL";
export const EDIT_MODAL = "EDIT_MODAL";
export const DELETE_MODAL = "DELETE_MODAL";
export const GET_FILMS_ASYNC = "GET_FILMS_ASYNC";
export const SET_FILTER = "SET_FILTER";
export const SET_GENRES = "SET_GENRES";
export const FILM_REQUEST_FAILURE = "FILM_REQUEST_FAILURE";
