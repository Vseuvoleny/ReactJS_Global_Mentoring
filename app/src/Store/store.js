export const initialState = {
  films: [],
  sortType: "release_date",
  error: "",
  genres: "All",
  film: undefined,
};
