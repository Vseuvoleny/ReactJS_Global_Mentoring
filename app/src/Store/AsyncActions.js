import axios from "axios";
import { BASE_URL } from "../utils/const/const";
import { getFilms, getErrorFromFetch } from "./ActionCreator";

export const asyncFetchFilms = (params) => {
  return (dispatch) => {
    axios
      .get(applyURL(params))
      .then((res) => {
        const { data } = res.data;
        dispatch(getFilms(data));
      })
      .catch((e) => {
        console.warn(e.message ?? e);
        dispatch(getErrorFromFetch(e.message ?? e));
      });
  };
};

const applyURL = ({ queryParams, genres, sortBy }) => {
  const params = queryParams ? `search=${queryParams}&searchBy=title` : "";
  const genre = genres ? `&filter=${genres}` : "";
  const sorting = sortBy ? `&sortBy=${sortBy}` : "";
  return `${BASE_URL}/movies?${params}${genre}${sorting}&sortOrder=desc&limit=6`;
};
