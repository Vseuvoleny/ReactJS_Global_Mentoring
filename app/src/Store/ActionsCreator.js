import axios from "axios";
import { GET_FILMS_BY_RELEASE_DATE } from "./ActionTypes";
import { BASE_URL } from "../utils/const/const";

export function fetchFilms() {
  return (dispatch) => {
    return axios
      .get(`${BASE_URL}/movies?sortBy=release_date&sortOrder=desc&limit=6`)
      .then((res) => {
        const { data } = res.data;
        dispatch(getFilmsByRelease(data));
      })
      .catch((e) => console.log(e.message ?? e));
  };
}

export function getFilmsByRelease(data) {
  return {
    type: GET_FILMS_BY_RELEASE_DATE,
    payload: data,
  };
}
