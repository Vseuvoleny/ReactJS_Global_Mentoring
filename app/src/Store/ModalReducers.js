import { APPLY_MODAL, DELETE_MODAL, EDIT_MODAL } from "./ActionTypes";

const initialState = {
  activeModal: "",
};

export const modalReducer = (state = initialState, action) => {
  switch (action.type) {
    case APPLY_MODAL:
      return { ...state, activeModal: APPLY_MODAL };
    case EDIT_MODAL:
      return { ...state, activeModal: EDIT_MODAL };
    case DELETE_MODAL:
      return { ...state, activeModal: DELETE_MODAL };
    default:
      return state;
  }
};
