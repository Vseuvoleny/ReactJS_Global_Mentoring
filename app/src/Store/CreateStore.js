import { createStore, applyMiddleware, combineReducers } from "redux";
import thunk from "redux-thunk";
import { initialState } from "./store";
import {
  GET_FILMS_ASYNC,
  SET_FILTER,
  SET_GENRES,
  FILM_REQUEST_FAILURE,
} from "./ActionTypes";
import { modalReducer } from "./ModalReducers";

function filmReducer(state = initialState, action) {
  switch (action.type) {
    case GET_FILMS_ASYNC:
      return {
        ...state,
        films: action.payload,
      };
    case SET_FILTER:
      return {
        ...state,
        sortType: action.payload,
      };
    case FILM_REQUEST_FAILURE:
      return {
        ...state,
        error: action.payload.error,
      };
    case SET_GENRES:
      return {
        ...state,
        genres: action.payload,
      };
    default:
      return state;
  }
}
const RootReducer = combineReducers({
  filmsStore: filmReducer,
  modalStore: modalReducer,
});

export const store = createStore(RootReducer, applyMiddleware(thunk));
