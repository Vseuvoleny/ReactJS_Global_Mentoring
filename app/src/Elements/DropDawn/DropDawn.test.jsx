import React from "react";
import { createStore } from "redux";
import { modalReducer } from "../../Store/ModalReducers";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import DropDawn from "./DropDawn";

const renderWithRedux = (
  component,
  { initialState, store = createStore(modalReducer, initialState) } = {}
) => {
  return {
    ...render(<Provider store={store}>{component}</Provider>),
    store,
  };
};

describe("Dropdawn element", () => {
  test("should render componenent", () => {
    const handleClick = jest.fn();
    const { getByText, container } = renderWithRedux(
      <DropDawn setDropDawnOpen={handleClick} />
    );
    expect(getByText(/Edit/i)).toBeInTheDocument();
    expect(getByText(/Delete/i)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
  test("should change type modal", () => {
    const initialState = {
      initialState: {
        activeModal: "",
      },
    };
    const typeEdit = jest.fn(
      () => (initialState.initialState.activeModal = "EDIT_MODAL")
    );
    const typeDelete = jest.fn(
      () => (initialState.initialState.activeModal = "DELETE_MODAL")
    );
    typeDelete();
    expect(initialState.initialState.activeModal).toBe("DELETE_MODAL");
    typeEdit();
    expect(initialState.initialState.activeModal).toBe("EDIT_MODAL");
  });
});
