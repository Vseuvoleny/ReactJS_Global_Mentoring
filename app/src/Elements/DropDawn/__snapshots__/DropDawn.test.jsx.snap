// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdawn element should render componenent 1`] = `
<div>
  <div
    class="dropdawn__container"
  >
    <svg
      class="close_icon"
      height="13"
      role="img"
      viewBox="0 0 12 13"
      width="12"
    >
      <path
        d="M1.09677 1.59464L10.9029 11.6236"
        stroke="white"
        stroke-linecap="round"
      />
      <path
        d="M10.9032 1.59464L1.0971 11.6236"
        stroke="white"
        stroke-linecap="round"
      />
    </svg>
    <div
      class="control"
    >
      Edit
    </div>
    <div
      class="control"
    >
      Delete
    </div>
  </div>
</div>
`;
